### ISSUE TITLE
Marco+Compton misses windows borders

### ISSUE THREAD

**Issue by dimka-rs** [mate-desktop/marco#533]
#### Expected behaviour
Windows have borders and can be resized with mouse.

#### Actual behaviour
Window borders are missing with marco + compton, available with marco + compositing.
Screenshots of window (alt+prtscn) are attached.

compton screenshot shows empty/transparent area where compositing screenshot shows underlying windows.
![compton](https://user-images.githubusercontent.com/295331/62424099-e9002900-b6d2-11e9-81be-61c65b432009.png)
![compositing](https://user-images.githubusercontent.com/295331/62424100-e9002900-b6d2-11e9-9d88-0f767c9947d8.png)

#### Steps to reproduce the behaviour
Upgrade 19.0 -> 19.1 -> 19.2

#### MATE general version
Host: dmk-mint Kernel: 4.15.0-55-generic x86_64 bits: 64 compiler: gcc v: 7.4.0 
Desktop: MATE 1.22.0 wm: marco dm: LightDM Distro: Linux Mint 19.2 Tina 
base: Ubuntu 18.04 bionic

#### Package version
Package: marco
Version: 1.22.2-1+tina

#### Linux Distribution
DISTRIB_ID=LinuxMint
DISTRIB_RELEASE=19.2
DISTRIB_CODENAME=tina
DISTRIB_DESCRIPTION="Linux Mint 19.2 Tina"

**Comment by lukefromdc** [comment]
This might be an issue in compton and not marco, I don't have compton installed (just marco and compiz-reloaded plus ICEWM and metacity available on my setup) and I have not had this happen. Note that the screenshot shows a known mate-screenshot bug 
https://github.com/mate-desktop/mate-utils/issues/236
in which screenshots include the new invisible resize area. That area is supposed to be usable to resize windows in themes that give a zero border-width. Wonder if that's not working with Compton, perhaps a communication issue with the compositor?

**Comment by dimka-rs** [comment]
@lukefromdc I've tried use pure Marco, but the borders are still missing, only Marco+Compositing works. Then I've tried to change themes and seems like only Mint-Y/Mint-Y-Dark/Mint-Y-Darker are affected. So this is theme related. Where should I report it?

**Comment by lukefromdc** [comment]
We do not control those themes, probably it's a problem with their metacity themes. Report it to Linuxmint's bugtracker.

**Comment by dimka-rs** [comment]
Thank you!

**Comment by monsta** [comment]
@lukefromdc: isn't https://github.com/mate-desktop/mate-themes/issues/249 related to this...?

**Comment by dimka-rs** [comment]
@monsta JFYI I've opened issue in mint-themes: https://github.com/linuxmint/mint-themes/issues/224

**Comment by lukefromdc** [comment]
That I do not know, I don't have the Mint themes installed, so I don't know what those themes are doing wrong. 
https://github.com/mate-desktop/marco/pull/503 in marco master and 
https://github.com/mate-desktop/marco/pull/518 in marco 1.22 fix that issue, in which decorations (window borders and titlebars) are shown but resizing windows is difficult if the theme sets a 1 or 2 px border and impossible if a theme sets a border width of zero.  These commits do NOT change (or are not supposed to change) what is visibly rendered on screen, they just allow you to resize a window from outside the visible border in the shadow area if compositing(necessary to have a shadow area that does not render solid black) is in use.

**Comment by monsta** [comment]
Yes, the problem is that these fixes apparently don't work w/o Marco's compositing...
(using Compton as an "external compositor" doesn't help, as reported here)

**Comment by lukefromdc** [comment]
Any way marco could explicity check for compton and then treat that as a compositing case? I've never used it (I have compiz here) but it's supposed to work with many window managers.

**Comment by vkareh** [comment]
@lukefromdc - huh, never actually thought of that, but the compositor code is setup in a plugin-like architecture. Xrender is just the only plugin, but no reason why we couldn't add others... Good idea, will give it some thought :)

**Comment by monsta** [comment]
That still leaves one case where Marco is without any compositor at all...

**Comment by lukefromdc** [comment]
How do those themes behave with marco set not to use compositing? BlackMATE (border width 1px) works fine and renders the borders, though it is very touchy aligning the mouse with those borders to resize (thus the BlackMATE-border variation that fixes that).

**Comment by monsta** [comment]
Same as reported here. No resize cursor is shown at all. Windows can be resized only via their titlebar. You can check it with Arc theme, for example.

**Comment by monsta** [comment]
I guess it started with metacity-theme-3 support addition.

**Comment by monsta** [comment]
Just tested with Metacity - apparently it adds 1px border to borderless (?) windows with Arc theme, so it's possible to resize them if I turn Metacity's compositor off.

**Comment by lukefromdc** [comment]
OK, that makes it clear what is going on: those themes set a zero-width border, which with the noncompositing codepath in marco gives a zero px width the mouse must be aligned with, therefore no possible alignment. Finding metacity's commit to add 1px to borderless windows and applying it so it only is used when marco's internal compositing is not used would fix this, as would modifying the offending themes (were they meant for Cinnamon maybe?) to include a border.

**Comment by monsta** [comment]
I don't think Arc theme does that for Cinnamon; probably just a design decision. Anyway, we can't control all such themes out there, so we'll have to use a workaround. Either we do the same as Metacity, or try to load only metacity-theme-2 when compositor is off (no idea if it's technically possible though).
